icefoganalytics/internal-data-portal
# Build Dataset Search Feature

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/42

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)

# Context

All Datasets “search” view - /datasets?where[...]

This wireframe suggests that the “search” field in the all datasets page is not actually a magic search, but in fact an expandable search panel. This search panel would have a list of filters you could search by.

Alternatively, build out a magic search that offers a single field and sends it to /datasets/search/:searchToken, and does either a bunch of basic ilike + or, or a true fuzzy search.

## Wireframes

Before search
![Image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/5a1c375c-9020-41f1-a48a-7673e3d41a74)

After search
![Image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/9fa0b902-0b20-4bd9-96ca-9250005972a7)

# Implementation

Use ugly database search instead of fancy fuzzy search (for now). In the future we should definitely upgrade to [Full Text Search](https://learn.microsoft.com/en-us/sql/relational-databases/search/full-text-search?view=sql-server-ver16).
Use a combination of scopes to cause search effect; probably will not scale well and will become slow. This will likely not be a problem until we have 1000+ dataset records.

# Screenshots

Advanced Filters open and all filters in use
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/44477727-1c49-4a67-a868-7aa5d8c1d762)

Filters cleared
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/73c93a98-30e1-4792-a271-351a9572f82e)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "All Datasets" page at http://localhost:8080/datasets?perPage=10&page=1
5. Check that you can search by:
    - dataset name.
    - dataset description
    - keywords
    - owner acronym segment
    > Note you can't currently search by deep acronym. e.g. ED-CS does not work. but CS or ED does.
6. Check that the search term is stored in the url and activated again on page load.
7. Check that "Advanced filters" work and are stored in the URL
8. Check that "Advanced filters" open/closed state is store in URL.
